### Activity 1 - Becoming a Data Scientist - Find and Explore a Dataset

Dataset I chose is on Pima Indians with Diabetes.  Dataset can be found here: https://archive.ics.uci.edu/ml/datasets/Pima+Indians+Diabetes

In [18]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('pima_indians_diabetes.csv')

In [3]:
df.head()

times_pregnant  plasma_glucose_concentration  diastolic_blood_pressure  \
0               6                           148                        72   
1               1                            85                        66   
2               8                           183                        64   
3               1                            89                        66   
4               0                           137                        40   

   tricepts_skin_fold_thickness  2_hour_serum_insulin  body_mass_index  \
0                            35                     0             33.6   
1                            29                     0             26.6   
2                             0                     0             23.3   
3                            23                    94             28.1   
4                            35                   168             43.1   

   diabetes_pedigree_function  age  class_variable  
0                       0.627   50               1  
1                       0.351   31               0  
2                       0.672   32               1  
3                       0.167   21               0  
4                       2.288   33               1

In [4]:
df.shape

(768, 9)

There are 768 rows and 9 columns.  Let's look at what the columns are now.

In [6]:
df.columns.values

array(['times_pregnant', 'plasma_glucose_concentration',
       'diastolic_blood_pressure', 'tricepts_skin_fold_thickness',
       '2_hour_serum_insulin', 'body_mass_index',
       'diabetes_pedigree_function', 'age', 'class_variable'], dtype=object)

In [11]:
df.dtypes

times_pregnant                    int64
plasma_glucose_concentration      int64
diastolic_blood_pressure          int64
tricepts_skin_fold_thickness      int64
2_hour_serum_insulin              int64
body_mass_index                 float64
diabetes_pedigree_function      float64
age                               int64
class_variable                    int64
dtype: object

In [13]:
df.describe()

times_pregnant  plasma_glucose_concentration  diastolic_blood_pressure  \
count      768.000000                    768.000000                768.000000   
mean         3.845052                    120.894531                 69.105469   
std          3.369578                     31.972618                 19.355807   
min          0.000000                      0.000000                  0.000000   
25%          1.000000                     99.000000                 62.000000   
50%          3.000000                    117.000000                 72.000000   
75%          6.000000                    140.250000                 80.000000   
max         17.000000                    199.000000                122.000000   

       tricepts_skin_fold_thickness  2_hour_serum_insulin  body_mass_index  \
count                    768.000000            768.000000       768.000000   
mean                      20.536458             79.799479        31.992578   
std                       15.952218            115.244002         7.884160   
min                        0.000000              0.000000         0.000000   
25%                        0.000000              0.000000        27.300000   
50%                       23.000000             30.500000        32.000000   
75%                       32.000000            127.250000        36.600000   
max                       99.000000            846.000000        67.100000   

       diabetes_pedigree_function         age  class_variable  
count                  768.000000  768.000000      768.000000  
mean                     0.471876   33.240885        0.348958  
std                      0.331329   11.760232        0.476951  
min                      0.078000   21.000000        0.000000  
25%                      0.243750   24.000000        0.000000  
50%                      0.372500   29.000000        0.000000  
75%                      0.626250   41.000000        1.000000  
max                      2.420000   81.000000        1.000000

In [20]:
df.isnull().sum()

times_pregnant                  0
plasma_glucose_concentration    0
diastolic_blood_pressure        0
tricepts_skin_fold_thickness    0
2_hour_serum_insulin            0
body_mass_index                 0
diabetes_pedigree_function      0
age                             0
class_variable                  0
dtype: int64

According to the description of the dataset there are missing values, but there are not showing any missing values.  I believe that is because the missing values are listed as "0"'s.  I'm going to fill these 0's with NaN.  I will not change times_pregnant and class variable because 0 is a valid answer for both of them

In [26]:
df['plasma_glucose_concentration'] = df['plasma_glucose_concentration'].replace('0', np.nan)
df['diastolic_blood_pressure'] = df['diastolic_blood_pressure'].replace('0', np.nan) 
df['tricepts_skin_fold_thickness'] = df['tricepts_skin_fold_thickness'].replace('0', np.nan) 
df['2_hour_serum_insulin'] = df['2_hour_serum_insulin'].replace('0', np.nan)        
df['body_mass_index'] = df['body_mass_index'].replace('0', np.nan) 
df['diabetes_pedigree_function'] = df['diabetes_pedigree_function'].replace('0', np.nan) 
df['age'] = df['age'].replace('0', np.nan) 

In [27]:
df.head()

times_pregnant  plasma_glucose_concentration  diastolic_blood_pressure  \
0               6                         148.0                      72.0   
1               1                          85.0                      66.0   
2               8                         183.0                      64.0   
3               1                          89.0                      66.0   
4               0                         137.0                      40.0   

   tricepts_skin_fold_thickness  2_hour_serum_insulin  body_mass_index  \
0                          35.0                   NaN             33.6   
1                          29.0                   NaN             26.6   
2                           NaN                   NaN             23.3   
3                          23.0                  94.0             28.1   
4                          35.0                 168.0             43.1   

   diabetes_pedigree_function  age  class_variable  
0                       0.627   50               1  
1                       0.351   31               0  
2                       0.672   32               1  
3                       0.167   21               0  
4                       2.288   33               1

In [28]:
df.isnull().sum()

times_pregnant                    0
plasma_glucose_concentration      5
diastolic_blood_pressure         35
tricepts_skin_fold_thickness    227
2_hour_serum_insulin            374
body_mass_index                  11
diabetes_pedigree_function        0
age                               0
class_variable                    0
dtype: int64

Now we have some missing values!  Lots of missing values in 2 Hour Serum Insulin and Tricepts Skin Fold Thickness.  Now I want to see if there are any rows that are missing a signifigant amount of data

In [29]:
df.apply(lambda x: x.count(), axis=1).sort_values(ascending=True)

706    5
49     5
522    5
426    5
60     5
494    5
81     5
535    6
193    6
336    6
222    6
269    6
468    6
697    6
261    6
78     6
484    6
703    6
266    6
300    6
643    6
332    6
347    6
619    6
601    6
430    6
589    6
7      6
435    6
9      6
      ..
234    9
449    9
220    9
465    9
466    9
199    9
493    9
203    9
204    9
490    9
206    9
487    9
486    9
485    9
208    9
483    9
482    9
480    9
478    9
477    9
476    9
213    9
214    9
215    9
216    9
469    9
217    9
467    9
498    9
383    9
dtype: int64

I have 7 rows missing 4 features.  

Now want to look at how many of each class value I have

In [16]:
df['class_variable'].sum()

268

In my dataset I have 268 with diabetes (class value of 1) and 500 without diabetes (class value of 0)